api/src/api/notebooks.ts

```diff
@@ -381,64 +383,139 @@ api.get(
   }
 );
 
-// export current versions of all records in this notebook as csv
+// Types for download format and token payloads
+const DownloadFormatSchema = z.enum(['csv', 'zip']);
+const DownloadTokenPayloadSchema = z.object({
+  projectID: z.string(),
+  format: DownloadFormatSchema,
+  viewID: z.string(),
+  userID: z.string(),
+});
+type DownloadTokenPayload = z.infer<typeof DownloadTokenPayloadSchema>;
+
+// download tokens last this long
+const DOWNLOAD_TOKEN_EXPIRY_MINUTES = 5;
+
+const generateDownloadToken = async ({
+  user,
+  payload,
+}: {
+  user: Express.User;
+  payload: DownloadTokenPayload;
+}) => {
+  const signingKey = await KEY_SERVICE.getSigningKey();
+  const token = await new SignJWT(payload)
+    .setProtectedHeader({
+      alg: signingKey.alg,
+      kid: signingKey.kid,
+    })
+    .setSubject(user.user_id)
+    .setIssuedAt()
+    .setIssuer(signingKey.instanceName)
+    // Expiry in minutes
+    .setExpirationTime(DOWNLOAD_TOKEN_EXPIRY_MINUTES.toString() + 'm')
+    .sign(signingKey.privateKey);
+  return token;
+};
+
+const validateDownloadToken = async ({
+  token,
+}: {
+  token: string;
+}): Promise<DownloadTokenPayload | null> => {
+  const signingKey = await KEY_SERVICE.getSigningKey();
+  try {
+    const result = await jwtVerify(token, signingKey.publicKey, {
+      algorithms: [signingKey.alg],
+      // verify issuer
+      issuer: signingKey.instanceName,
+    });
+    return result.payload as DownloadTokenPayload;
```

Comment:
You have the zod model here - you could also parse it to make sure it's valid since this is a security sensitive action.